File: src/main/java/de/md5lukas/commands/SubCommand.java
package de.md5lukas.commands;

import org.bukkit.command.CommandSender;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

public abstract class SubCommand {

    private List<SubCommand> subCommands;

    public SubCommand() {
        subCommands = new ArrayList<>();
    }

    public boolean guard(CommandSender sender, List<String> args) {
        return true;
    }

    public SubCommand addSubCommand(SubCommand command) {
        subCommands.add(command);
        return this;
    }

    public abstract String getCommand();

    public List<String> getAliases() {
        return Collections.emptyList();
    }

    //<editor-fold desc="onCommand">
    public final boolean onCommand(CommandSender sender, LinkedList<String> args) {
        if (!guard(sender, args)) {
            return false;
        }
        if (args.isEmpty()) {
            return run(sender);
        }
        String subCommandString = args.getFirst().toLowerCase();
        for (SubCommand subCommand : subCommands) {
            if (subCommand.getCommand().equals(subCommandString) || subCommand.getAliases().contains(subCommandString)) {
                return subCommand.onCommand(sender, args);
            }
        }
        return run(sender, args);
    }

    public boolean run(CommandSender sender) {
        return false;
    }

    public boolean run(CommandSender sender, List<String> args) {
        return false;
    }
    //</editor-fold>

    //<editor-fold desc="onTabComplete">
    public final List<String> onTabComplete(CommandSender sender, LinkedList<String> args) {
        if (!guard(sender, args)) {
            return Collections.emptyList();
        }
        if (args.isEmpty()) {
            return tabComplete(sender);
        }
        String subCommandString = args.getFirst().toLowerCase();
        for (SubCommand subCommand : subCommands) {
            if (subCommand.getCommand().equals(subCommandString) || subCommand.getAliases().contains(subCommandString)) {
                return subCommand.onTabComplete(sender, args);
            }
        }
        List<String> completions = subCommands.stream().map(SubCommand::getCommand).collect(Collectors.toList());
        completions.addAll(tabComplete(sender, args));
        return completions;
    }

    public List<String> tabComplete(CommandSender sender) {
        return Collections.emptyList();
    }

    public List<String> tabComplete(CommandSender sender, List<String> args) {
        return Collections.emptyList();
    }
    //</editor-fold>
}

File: src/main/kotlin/de/md5lukas/commands/SubCommand.kt
package de.md5lukas.commands

import de.md5lukas.commands.dsl.options.BaseCommandOptions
import de.md5lukas.commands.dsl.options.SubCommandOptions
import de.md5lukas.commons.collections.PaginationList
import org.bukkit.command.CommandSender
import java.util.*

open class SubCommand(
    private val rootOptions: BaseCommandOptions,
    options: SubCommandOptions,
    fullCommand: String
) {

    private val fullCommand: String = fullCommand + options.name

    private val subCommands = options.subCommands.map { SubCommand(rootOptions, it, this.fullCommand) }

    protected val name = options.name
    private val aliases = options.aliases
    private val shortDescription = options.shortDescription
    private val description = options.description

    private val guard = options.guard
    private val guardFailed = options.guardFailed

    private val run = options.run
    private val runArgs = options.runArgs

    private val tabCompleter = options.tabCompleter

    private val helpEntries: PaginationList<((sender: CommandSender) -> Unit)> =
        PaginationList(rootOptions.helpListingsPerPage)

    init {
        helpEntries.add(::printHelp)
        subCommands.forEach {
            helpEntries.add(it::printShortHelp)
        }
    }

    protected fun onCommand(sender: CommandSender, args: LinkedList<String>, fullCommand: String) {
        if (!guard(sender)) {
            guardFailed(sender)
            return
        }
        if (args.isEmpty()) {
            if (run == null) {
                rootOptions.notFoundMessage(sender, fullCommand)
            } else {
                run.invoke(sender)
            }
            return
        }
        val subCommandName = args.first().toLowerCase()
        val newFullCommand = if (fullCommand.isEmpty()) {
            subCommandName
        } else {
            "$fullCommand $subCommandName"
        }
        for (subCommand in subCommands) {
            if (subCommand.nameMatches(subCommandName)) {
                args.removeFirst()
                subCommand.onCommand(
                    sender,
                    args,
                    newFullCommand
                )
                return
            }
        }
        if (runArgs == null) {
            rootOptions.notFoundMessage(sender, newFullCommand)
        } else {
            runArgs.invoke(sender, args)
        }
    }

    protected fun onTabComplete(sender: CommandSender, args: LinkedList<String>): Set<String> {
        if (!guard(sender)) {
            return emptySet()
        }
        when (args.size) {
            0 -> return emptySet()
            1 -> {
                return subCommands.map { it.name }.toMutableSet().also {
                    it.addAll(tabCompleter(sender))
                }
            }
            else -> {
                val subCommandName = args.removeFirstOrNull()?.toLowerCase()
                if (subCommandName != null) {
                    for (subCommand in subCommands) {
                        if (subCommand.nameMatches(subCommandName)) {
                            return subCommand.onTabComplete(sender, args)
                        }
                    }
                }
                return emptySet()
            }
        }
    }

    private fun nameMatches(string: String): Boolean {
        return name == string || aliases.contains(string)
    }

    private fun printHelp(sender: CommandSender) {
        rootOptions.helpFormatter(sender, name, fullCommand, description(sender))
    }

    private fun printShortHelp(sender: CommandSender) {
        rootOptions.shortHelpFormatter(sender, name, fullCommand, shortDescription(sender))
    }
}